File: mobile/lib/modules/cart/cart_actions.dart
import 'package:get/get.dart';

import '../../data/models/product.dart';
import '../../data/models/product_variant.dart';
import 'cart_service.dart';

Future<bool> addVariantToCart(
  Product product,
  ProductVariant variant, {
  int quantity = 1,
}) async {
  if (product.status.toUpperCase() != 'ACTIVE') {
    Get.snackbar(
      'cart_action.unavailable_title'.tr,
      'cart_action.unavailable_product'.tr,
      snackPosition: SnackPosition.BOTTOM,
    );
    return false;
  }

  if (!variant.inStock) {
    Get.snackbar(
      'cart_action.out_of_stock_title'.tr,
      'cart_action.out_of_stock_message'.tr,
      snackPosition: SnackPosition.BOTTOM,
    );
    return false;
  }

  final cart = Get.find<CartService>();
  final current = cart.quantityFor(variant.id);
  final requested = current + quantity;

  if (current > 0) {
    await cart.setQuantity(variant.id, requested);
  } else {
    await cart.add(variant.id, quantity: quantity);
  }
  return true;
}

Future<bool> addProductToCart(Product product, {int quantity = 1}) {
  final variant = product.firstInStockVariant ?? product.defaultVariant;
  if (variant == null) {
    Get.snackbar(
      'cart_action.no_options_title'.tr,
      'cart_action.no_options_message'.tr,
      snackPosition: SnackPosition.BOTTOM,
    );
    return Future.value(false);
  }
  return addVariantToCart(product, variant, quantity: quantity);
}

File: web/src/lib/cartActions.ts
import { useCartStore } from "@/stores/cartStore";
import type { Product, ProductVariant } from "@/types";

export type AddToCartResult =
  | { ok: true }
  | { ok: false; reason: "unavailable" | "out_of_stock" | "no_variant" };

export function addVariantToCart(
  product: Product,
  variant: ProductVariant,
  quantity = 1,
): AddToCartResult {
  if (product.status.toUpperCase() !== "ACTIVE") {
    return { ok: false, reason: "unavailable" };
  }

  if (!variant.inStock) {
    return { ok: false, reason: "out_of_stock" };
  }

  const cart = useCartStore.getState();
  const current = cart.quantityFor(variant.id);

  if (current > 0) {
    cart.setQuantity(variant.id, current + quantity);
  } else {
    cart.add(variant.id, quantity);
  }

  return { ok: true };
}

export function addToCartFailureMessage(
  reason: Exclude<AddToCartResult, { ok: true }>["reason"],
  t: (key: string) => string,
): string {
  switch (reason) {
    case "unavailable":
      return t("cart_action.unavailable_product");
    case "out_of_stock":
      return t("cart_action.out_of_stock_message");
    case "no_variant":
      return t("cart_action.no_options_message");
  }
}
